runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDos.java

```diff
@@ -0,0 +1,348 @@
+  }
+
+  // TODO: Implement
+  private static StateSpec toProto(StateDeclaration state) {
+    return StateSpec.newBuilder().build();
+  }
+
+  // TODO: Implement
+  private static TimerSpec toProto(TimerDeclaration timer) {
+    return TimerSpec.newBuilder().build();
+  }
+
+  @AutoValue
+  abstract static class DoFnAndMainOutput implements Serializable {
+    public static DoFnAndMainOutput of(
+        DoFn<?, ?> fn, TupleTag<?> tag) {
+      return new AutoValue_ParDos_DoFnAndMainOutput(fn, tag);
+    }
+
+    abstract DoFn<?, ?> getDoFn();
+    abstract TupleTag<?> getMainOutputTag();
+  }
+
+  private static SdkFunctionSpec toProto(DoFn<?, ?> fn, TupleTag<?> tag) {
+    return SdkFunctionSpec.newBuilder()
+        .setSpec(
+            FunctionSpec.newBuilder()
+                .setUrn(CUSTOM_JAVA_DO_FN_URN)
+                .setParameter(
+                    Any.pack(
+                        BytesValue.newBuilder()
+                            .setValue(
+                                ByteString.copyFrom(
+                                    SerializableUtils.serializeToByteArray(
+                                        DoFnAndMainOutput.of(fn, tag))))
+                            .build())))
+        .build();
+  }
+
+  private static DoFnAndMainOutput doFnAndMainOutputTagFromProto(SdkFunctionSpec fnSpec)
+      throws InvalidProtocolBufferException {
+    checkArgument(fnSpec.getSpec().getUrn().equals(CUSTOM_JAVA_DO_FN_URN));
+    byte[] serializedFn =
+        fnSpec.getSpec().getParameter().unpack(BytesValue.class).getValue().toByteArray();
+    return (DoFnAndMainOutput)
+        SerializableUtils.deserializeFromByteArray(serializedFn, "Custom DoFn And Main Output tag");
+  }
+
+  private static Optional<RunnerApi.Parameter> toProto(Parameter parameter) {
+    Cases<Type> runnerApiCases =
+        new Cases<Type>() {
+          @Override
+          public Type dispatch(StartBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(FinishBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(ProcessContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(OnTimerContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(WindowParameter p) {
+            return Type.WINDOW;
+          }
+
+          @Override
+          public Type dispatch(RestrictionTrackerParameter p) {
+            return Type.RESTRICTION_TRACKER;
+          }
+
+          @Override
+          public Type dispatch(StateParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(TimerParameter p) {
+            return null;
+          }
+        };
+    Type parameterType = parameter.match(runnerApiCases);
```

Comment:
(nit) I prefer putting the `new Cases` inline so it reads like the enhanced `switch` that it is.